File: clients/typescript/src/model/flying-worker-group.ts
import { Resource } from "./resource";
import { Stream } from "../stream";

/**
 * A flying worker group
 */
export class FlyingWorkerGroup {
    /**
     * Index of player controlling workers
     */
    playerIndex: number
    /**
     * Number of workers in the group
     */
    number: number
    /**
     * Tick when workers left previous planet on their path
     */
    departureTick: number
    /**
     * Id of the previous planet on the path
     */
    departurePlanet: number
    /**
     * Tick when workers will arrive to the next planet in their path
     */
    nextPlanetArrivalTick: number
    /**
     * Id of the next planet in the path
     */
    nextPlanet: number
    /**
     * Id of the target planet
     */
    targetPlanet: number
    /**
     * Resource that workers are carrying
     */
    resource: Resource | null

    constructor(playerIndex: number, number: number, departureTick: number, departurePlanet: number, nextPlanetArrivalTick: number, nextPlanet: number, targetPlanet: number, resource: Resource | null) {
        this.playerIndex = playerIndex;
        this.number = number;
        this.departureTick = departureTick;
        this.departurePlanet = departurePlanet;
        this.nextPlanetArrivalTick = nextPlanetArrivalTick;
        this.nextPlanet = nextPlanet;
        this.targetPlanet = targetPlanet;
        this.resource = resource;
    }

    /**
     * Read FlyingWorkerGroup from input stream
     */
    static async readFrom(stream: Stream): Promise<FlyingWorkerGroup> {
        let playerIndex;
        playerIndex = await stream.readInt();
        let number;
        number = await stream.readInt();
        let departureTick;
        departureTick = await stream.readInt();
        let departurePlanet;
        departurePlanet = await stream.readInt();
        let nextPlanetArrivalTick;
        nextPlanetArrivalTick = await stream.readInt();
        let nextPlanet;
        nextPlanet = await stream.readInt();
        let targetPlanet;
        targetPlanet = await stream.readInt();
        let resource;
        if (await stream.readBool()) {
            resource = await Resource.readFrom(stream);
        } else {
            resource = null;
        }
        return new FlyingWorkerGroup(playerIndex, number, departureTick, departurePlanet, nextPlanetArrivalTick, nextPlanet, targetPlanet, resource)
    }

    /**
     * Write FlyingWorkerGroup to output stream
     */
    async writeTo(stream: Stream) {
        let playerIndex = this.playerIndex;
        await stream.writeInt(playerIndex);
        let number = this.number;
        await stream.writeInt(number);
        let departureTick = this.departureTick;
        await stream.writeInt(departureTick);
        let departurePlanet = this.departurePlanet;
        await stream.writeInt(departurePlanet);
        let nextPlanetArrivalTick = this.nextPlanetArrivalTick;
        await stream.writeInt(nextPlanetArrivalTick);
        let nextPlanet = this.nextPlanet;
        await stream.writeInt(nextPlanet);
        let targetPlanet = this.targetPlanet;
        await stream.writeInt(targetPlanet);
        let resource = this.resource;
        if (resource === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await resource.writeTo(stream);
        }
    }
}

File: clients/python/model/flying_worker_group.py
from model.resource import Resource
from stream_wrapper import StreamWrapper
from typing import Optional

class FlyingWorkerGroup:
    """A flying worker group"""

    __slots__ = ("player_index","number","departure_tick","departure_planet","next_planet_arrival_tick","next_planet","target_planet","resource",)

    player_index: int
    number: int
    departure_tick: int
    departure_planet: int
    next_planet_arrival_tick: int
    next_planet: int
    target_planet: int
    resource: Optional[Resource]

    def __init__(self, player_index: int, number: int, departure_tick: int, departure_planet: int, next_planet_arrival_tick: int, next_planet: int, target_planet: int, resource: Optional[Resource]):
        self.player_index = player_index
        """Index of player controlling workers"""
        self.number = number
        """Number of workers in the group"""
        self.departure_tick = departure_tick
        """Tick when workers left previous planet on their path"""
        self.departure_planet = departure_planet
        """Id of the previous planet on the path"""
        self.next_planet_arrival_tick = next_planet_arrival_tick
        """Tick when workers will arrive to the next planet in their path"""
        self.next_planet = next_planet
        """Id of the next planet in the path"""
        self.target_planet = target_planet
        """Id of the target planet"""
        self.resource = resource
        """Resource that workers are carrying"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "FlyingWorkerGroup":
        """Read FlyingWorkerGroup from input stream
        """
        player_index = stream.read_int()
        number = stream.read_int()
        departure_tick = stream.read_int()
        departure_planet = stream.read_int()
        next_planet_arrival_tick = stream.read_int()
        next_planet = stream.read_int()
        target_planet = stream.read_int()
        if stream.read_bool():
            resource = Resource(stream.read_int())
        else:
            resource = None
        return FlyingWorkerGroup(player_index, number, departure_tick, departure_planet, next_planet_arrival_tick, next_planet, target_planet, resource)
    
    def write_to(self, stream: StreamWrapper):
        """Write FlyingWorkerGroup to output stream
        """
        stream.write_int(self.player_index)
        stream.write_int(self.number)
        stream.write_int(self.departure_tick)
        stream.write_int(self.departure_planet)
        stream.write_int(self.next_planet_arrival_tick)
        stream.write_int(self.next_planet)
        stream.write_int(self.target_planet)
        if self.resource is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.resource)
    
    def __repr__(self):
        return "FlyingWorkerGroup(" + \
            repr(self.player_index) + \
            ", " + \
            repr(self.number) + \
            ", " + \
            repr(self.departure_tick) + \
            ", " + \
            repr(self.departure_planet) + \
            ", " + \
            repr(self.next_planet_arrival_tick) + \
            ", " + \
            repr(self.next_planet) + \
            ", " + \
            repr(self.target_planet) + \
            ", " + \
            repr(self.resource) + \
            ")"